### ISSUE TITLE
errors: want better error formatting in Sentry report

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#52600]
I wanna use the example Sentry report in #52550 to air some gripes about our error formatting.
I don't necessarily have very coherent suggestions, but anything you can do to help me Rafa...

The report in question is https://sentry.io/organizations/cockroach-labs/issues/1832385016/?project=164528&referrer=webhooks_plugin#packages

```
Message:

*errors.errorString
*safedetails.withSafeDetails: %s: %+v (1)
store_raft.go:494: *withstack.withStack (top exception)
(check the extra data payloads)

Additional data:

%s: %+v 
-- arg 1: while committing batch 
-- arg 2: replica_raft.go:707: <*storage.Error> 
wrapper: <*errutil.withMessage> 
wrapper: <*withstack.withStack> 
(more details:) 
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked 
/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:707 github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady 
/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:425 github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady 	
/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:493 github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker 	
/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:226 github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2 	
/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:166 github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1 	
/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:222 runtime.goexit 	/usr/local/go/src/runtime/asm_amd64.s:1373
```

It takes a lot of staring to parse these snippets. First off, the Sentry `Message:` has almost no information. It's not even clear that this is a `log.Fatal()`. The only thing that it has is the location of the `log.Fatal()` call: `store_raft.go:494`; the rest is gibberish to me. It would have been significantly better if I only saw that and nothing else. But ideally I'd see more than that, namely some of the below.

Then, the `Additional data` starts with:
```
%s: %+v 
```
This seems pretty useless. It corresponds to this line:
```
log.FatalfDepth(ctx, 1, "%s: %+v", log.Safe(expl), err)
```
It would have been much better if the first line was something like:
```
while committing batch: <*storage.Error>
```

Then the next lines are:
```
-- arg 1: while committing batch 
-- arg 2: replica_raft.go:707: <*storage.Error> 
wrapper: <*errutil.withMessage> 
wrapper: <*withstack.withStack> 
(more details:) 
<stack>
```
When I saw that, my first question was "arguments to *what*"? The answer (to the `log.Fatalf()` call) and the fact that these two things corresponded to the `%s` and the `%v` from above, respectively, was not clear to me.

If the first arg would have been included in the message, then perhaps we could have gotten rid of the `--- arg 1` line.
For `-- arg 2`, there's the question about whether there's a missing message for `storage.Error` or not. 
Then, I think the line `wrapper: <*errutil.withMessage>` is pretty useless. What I would have liked to see is the message of that wrapper. By looking at the code, the message is "while committing batch" (the same as we've seen in `--arg 1`). The code jumps through hoops to pass that message both to the `log.Fatal()` call and to the `errors.Wrap()`. I don't know if the code is silly or if there's no better way to achieve what it wants (which presumably is to make sure that string makes it to these reports); I would have expected the message of `errors.Wrap()` to be considered safe (or, if not by default, perhaps there should be an `errors.WrapSafe()` or so).
Without a message, the `errutil.withMessage` line I think should go away. Similarly, I think the line `wrapper: <*withstack.withStack>` doesn't add anything; I'd get rid of it. And then I don't know what's up with the `(more details:)` line, but I'd definitely get rid of it :).

**Comment by knz** [comment]
> Then, the `Additional data` starts with:
> 
> ```
> %s: %+v 
> ```
> 
> This seems pretty useless. It corresponds to this line:
> 
> ```
> log.FatalfDepth(ctx, 1, "%s: %+v", log.Safe(expl), err)
> ```
> 
> It would have been much better if the first line was something like:
> 
> ```
> while committing batch: <*storage.Error>
> ```

Yes this has been annoying (note: this has always been this way ever since the first version of crash reporting).

I think this will be improved by introducing native support for redaction markers in the errors library.

> Then the next lines are:
> 
> ```
> -- arg 1: while committing batch 
> -- arg 2: replica_raft.go:707: <*storage.Error> 
> ```
> 
> When I saw that, my first question was "arguments to _what_"? The answer (to the `log.Fatalf()` call) and the fact that these two things corresponded to the `%s` and the `%v` from above, respectively, was not clear to me.

That needs to be documented. In the general case, we're not going to re-do the string splicing done by Printf during crash reporting, that's too late for that.


> For `-- arg 2`, there's the question about whether there's a missing message for `storage.Error` or not.

Yes I agree this is a bug https://github.com/cockroachdb/errors/issues/40

> Then, I think the line `wrapper: <*errutil.withMessage>` is pretty useless. What I would have liked to see is the message of that wrapper

The reason why you see `<*errutil.withMessage>` here is because the wrapper does not contain a safe string. If it did, you'd see that here instead of just the type.

> By looking at the code, the message is "while committing batch" (the same as we've seen in `--arg 1`).

I believe that's incorrect in this case (see above).

> The code jumps through hoops to pass that message both to the `log.Fatal()` call and to the `errors.Wrap()`.

Where do you see this?

> I would have expected the message of `errors.Wrap()` to be considered safe (or, if not by default, perhaps there should be an `errors.WrapSafe()` or so).

This is a legitimate ask, already tracked in #49752 and #49659. 

> Similarly, I think the line `wrapper: <*withstack.withStack>` doesn't add anything; I'd get rid of it.

There can be multiple types of error that contain stack traces. Having the type tells you more about where the error is coming from.

> And then I don't know what's up with the `(more details:)` line, but I'd definitely get rid of it :).

The situation you've missed throughout (and the current formatting does not help) is that the payload is formatting two different objects:

- the "main" object is the `log.Fatal` call itself. The main stack trace (under "Exception") is that of the log.Fatal call.  
- the entire string after the prefix `-- arg: 2` in the box "1: details" is the content of the storage `error` that was passed as argument to the `log.Fatal`. This entire string:

```
replica_raft.go:707: <*storage.Error>
wrapper: <*errutil.withMessage>
wrapper: <*withstack.withStack>
(more details:)
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:707
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:425
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:493
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:226
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:166
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:222
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1373
```

is a (safe) textual representation of the `error` argument.

Note that the stack trace "in the error" is a different one from where `log.Fatal`  was called.
If the error object had been created "far" from the log.Fatal call, they would be even more different.

I could see a reason to extract that information into a higher level in the Sentry report, but I think it is difficult. For the time being, I will improve the translation of `error` to string details to make it clear that you're looking at an error object in string form.

**Comment by andreimatei** [comment]
>> Then, the Additional data starts with:
>>%s: %+v 
>>This seems pretty useless. It corresponds to this line:
>>log.FatalfDepth(ctx, 1, "%s: %+v", log.Safe(expl), err)
>>It would have been much better if the first line was something like:
>>`while committing batch: <*storage.Error>`

> Yes this has been annoying (note: this has always been this way ever since the first version of crash reporting).
> I think this will be improved by introducing native support for redaction markers in the errors library.

I'm not sure what "introducing support for redaction markers" means exactly, but the way I imagined I'd get what I want is by recognizing that the first argument to `log.Fatal()` here is a `log.Safe()`, so we can just render it instead of the `%s`. No?

> The reason why you see <*errutil.withMessage> here is because the wrapper does not contain a safe string. If it did, you'd see that here instead of just the type.

Right; I think we really should address #49752 and #49659 soon; I think it'll make a big difference.

> > The code jumps through hoops to pass that message both to the log.Fatal() call and to the errors.Wrap().

> Where do you see this?

See [here](https://github.com/cockroachdb/cockroach/blob/4d29d15579928f544a37974ccbf080287c6adae1/pkg/kv/kvserver/replica_raft.go#L707) and a bunch of other returns around there. We're both returning an explanation like `while committing batch` and also using it to wrap the error that we also return. The only reason we return that explanation is so that we can stick it into a `log.Safe()`, together with the error. This suggests to me that we're working around some siliness.

> The situation you've missed throughout (and the current formatting does not help) is that the payload is formatting two different objects:

Yeah; one of my main complaints here is that the words `log.Fatal()` don't appear anywhere in that report. If they did (both in the `Message` and in the `Exception`), I think that's help a lot.

**Comment by knz** [comment]
> I'm not sure what "introducing support for redaction markers" means exactly, but the way I imagined I'd get what I want is by recognizing that the first argument to `log.Fatal()` here is a `log.Safe()`, so we can just render it instead of the `%s`. No?

Yes that is what the new `redact` package does. However it's not yet used for errors. I'm adding that.

> > The reason why you see <*errutil.withMessage> here is because the wrapper does not contain a safe string. If it did, you'd see that here instead of just the type.
> 
> Right; I think we really should address #49752 and #49659 soon; I think it'll make a big difference.

Agreed!

> > Where do you see this?
> 
> See [here](https://github.com/cockroachdb/cockroach/blob/4d29d15579928f544a37974ccbf080287c6adae1/pkg/kv/kvserver/replica_raft.go#L707) [...]

This code is indeed silly. There's no need for so much annotations.
Once the solution to  #49752 and #49659 goes in, we can log the error directly in the `log.Fatal` call and no need to re-pass the `expl` string.

> > The situation you've missed throughout (and the current formatting does not help) is that the payload is formatting two different objects:
> 
> Yeah; one of my main complaints here is that the words `log.Fatal()` don't appear anywhere in that report. If they did (both in the `Message` and in the `Exception`), I think that's help a lot.

That's another concern and you are very right. Will send a PR shortly.

**Comment by knz** [comment]
This has been addressed by the latest errors/redact code in 20.2.

The 20.1 and prior reports will have some improvements but not as dramatic.
